tedana/decomposition/gift_pca.py

```diff
@@ -0,0 +1,652 @@
+        vcu = np.zeros((2 * n[0] - 1, 2 * n[1] - 1, 2 * n[2] - 1))
+        for m3 in range(n[2]):
+            vcu[:, :, (n[2] - 1) - m3] = vd * v3[m3]
+            vcu[:, :, (n[2] - 1) + m3] = vd * v3[m3]
+
+        # Possible source of NAN values
+        xc = xc / vcu
+
+        # Scale Parzen windows
+        parzen_window_2D = np.dot(parzen_w_1[np.newaxis, :].T,
+                                  parzen_w_2[np.newaxis, :])
+        parzen_window_3D = np.zeros((2 * n[0] - 1, 2 * n[1] - 1, 2 * n[2] - 1))
+        for m3 in range(n[2] - 1):
+            parzen_window_3D[:, :, (n[2] - 1) - m3] = np.dot(
+                parzen_window_2D, parzen_w_3[n[2] - 1 - m3])
+            parzen_window_3D[:, :, (n[2] - 1) + m3] = np.dot(
+                parzen_window_2D, parzen_w_3[n[2] - 1 + m3])
+
+        # Apply 3D Parzen Window
+        xc = xc * parzen_window_3D
+        xf = fftshift(fftn(xc))
+
+    else:
+        raise ValueError('Unrecognized matrix dimension.')
+
+    xf = abs(xf)
+    xf[xf < 1e-4] = 1e-4
+
+    # Estimation of the entropy rate
+    out = 0.5 * np.log(2 * np.pi * np.exp(1)) + _sumN(np.log(abs(
+        (xf)))) / 2 / _sumN(abs(xf))
+
+    return out
+
+
+def _est_indp_sp(x):
+    """
+    Estimate the effective number of independent samples based on the maximum
+    entropy rate principle of stationary random process.
+
+    Parameters
+    ----------
+    x : ndarray
+        The data to have the number of samples estimated
+
+    Returns
+    -------
+    s : int
+        Number of iterations required to estimate entropy rate
+    entrate_m : float
+        The entropy rate of the data
+
+    Notes
+    -----
+    TOOD: explain math
+
+    References
+    ----------
+    TODO: add references
+    """
+
+    dimv = x.shape
+    s0 = None
+
+    for j in range(np.min(dimv) - 1):
+        x_sb = _subsampling(x, j + 1)
+        entrate_m = _entrate_sp(x_sb, 1)
+
+        ent_ref = 1.41
+        if entrate_m > ent_ref:
+            s0 = j
+            break
+
+    if not s0:
+        raise ValueError('Ill conditioned data, can not estimate'
+                         'independent samples.')
+    s = s0
+    LGR.debug('Estimated the entropy rate of the Gaussian component '
+              'with subsampling depth {}'.format(j))
+
+    return s, entrate_m
+
+
+def _subsampling(x, s):
+    """
+    Subsampling the data evenly with space 's'.
+
+    Parameters
+    ----------
+    x : ndarray
+        The data to be subsampled
+    s : int
+        The subsampling depth
+
+    Returns
+    -------
+    out : ndarray
+        Subsampled data
+    """
+
+    # First index from which to start subsampling for each dimension
+    x0 = [0, 0, 0]
+    n = x.shape
+
+    if x.ndim == 3 and np.min(n) != 1:  # 3D
+        out = x[np.arange(
+            x0[0], n[0], s), :, :][:, np.arange(
+                x0[1], n[1], s), :][:, :, np.arange(x0[2], n[2], s)]
+    else:
+        raise ValueError('Unrecognized matrix dimension!(subsampling)')
```

Comment:
I don't understand `(subsampling)` in this error message. Can we clarify or remove ?

```suggestion
        raise ValueError('Unrecognized matrix dimension!')
```